plutus-core/cost-model/create-cost-model/CreateBuiltinCostModel.hs

```diff
@@ -235,241 +228,217 @@ createBuiltinCostModel bmfile rfile = do
+    "quadratic_in_y"       -> ModelTwoArgumentsQuadraticInY       <$> readOneVariableQuadraticFunction "y_mem" e
+    _                      -> error $ "Unknown two-variable model type: " ++ ty
+
+loadThreeVariableCostingFunction :: MonadR m => SomeSEXP (Region m) -> m ModelThreeArguments
+loadThreeVariableCostingFunction e = do
+  ty <- getType e
+  case ty of
+    "constant_cost"               -> ModelThreeArgumentsConstantCost          <$> getConstant e
+    "linear_in_x"                 -> ModelThreeArgumentsLinearInX             <$> readOneVariableLinearFunction "x_mem" e
+    "linear_in_y"                 -> ModelThreeArgumentsLinearInY             <$> readOneVariableLinearFunction "y_mem" e
+    "linear_in_z"                 -> ModelThreeArgumentsLinearInZ             <$> readOneVariableLinearFunction "z_mem" e
+    "quadratic_in_z"              -> ModelThreeArgumentsQuadraticInZ          <$> readOneVariableQuadraticFunction "z_mem" e
+    "literal_in_y_or_linear_in_z" -> ModelThreeArgumentsLiteralInYOrLinearInZ <$> error "literal"
+    _                             -> error $ "Unknown three-variable model type: " ++ ty
+
+loadSixVariableCostingFunction :: MonadR m => SomeSEXP (Region m) -> m ModelSixArguments
+loadSixVariableCostingFunction e = do
+  ty <- getType e
+  case ty of
+    "constant_cost" -> ModelSixArgumentsConstantCost <$> getConstant e
+    _               -> error $ "Unknown six-variable model type: " ++ ty
 
 boolMemModel :: ModelTwoArguments
 boolMemModel = ModelTwoArgumentsConstantCost 1
 
 memoryUsageAsCostingInteger :: ExMemoryUsage a => a -> CostingInteger
 memoryUsageAsCostingInteger = coerce . sumCostStream . flattenCostRose . memoryUsage
 
+-- | The types of functions which take an R SEXP and extract a CPU model from
+-- it, then pair it up with a memory costing function.
+type MakeCostingFun1 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelOneArgument)
+type MakeCostingFun2 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelTwoArguments)
+type MakeCostingFun3 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelThreeArguments)
+type MakeCostingFun6 = forall m . MonadR m => SomeSEXP (Region m) -> m (CostingFun ModelSixArguments)
+
+-- | Create a
+makeOneVariableCostingFunction :: ModelOneArgument -> MakeCostingFun1
+makeOneVariableCostingFunction memModel e =
+  CostingFun <$> (loadOneVariableCostingFunction e) <*> pure memModel
+
+makeTwoVariableCostingFunction :: ModelTwoArguments -> MakeCostingFun2
+makeTwoVariableCostingFunction memModel e =
+  CostingFun <$> (loadTwoVariableCostingFunction e) <*> pure memModel
+
+makeThreeVariableCostingFunction :: ModelThreeArguments -> MakeCostingFun3
+makeThreeVariableCostingFunction memModel e =
+  CostingFun <$> (loadThreeVariableCostingFunction e) <*> pure memModel
+
+makeSixVariableCostingFunction :: ModelSixArguments -> MakeCostingFun6
+makeSixVariableCostingFunction memModel e =
```

Comment:
I'd make all these `loadOneVariableCostingFunction`, `loadTwoVariableCostingFunction` etc into a single method of a type class taking `ModelOneArgument`, `ModelTwoArguments` etc at the type-level.